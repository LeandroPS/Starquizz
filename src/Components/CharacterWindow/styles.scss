@import '../../styles/common';
@import '../../styles/colors';

.character-window{

    width: 600px;
    background: #fff;
    border-radius: 10px;
    overflow: hidden;

    position: fixed;
    top: calc(50% + 50px);
    left: 50%;
    transform: translate(-50%, -50%);

    opacity: 0;
    visibility: hidden;

    transition: all 0.3s;

    z-index: 2;

    @extend %box-shadow;

    @media screen and (max-width: 700px){
        width: 300px;
    }

    &.show{
        top: 50%;
        opacity: 1;
        visibility: visible;

        transition: opacity 0.3s, top 0.3s;
    }

    button.close{
        position: absolute;
        top: 10px;
        right: 10px;
    }

    .inner{
        display: flex;

        .thumbnail{
            width: 300px;
            height: 300px;
            display: inline-block;
            background-color: $gray;
            background-position: top center;
            background-size: cover;
            background-repeat: no-repeat;

            @media screen and (max-width: 700px){
                display: none;
            }
        }
    
        .details{
            width: 300px;
            display: inline-block;
            padding: 40px 20px 20px 20px;

            h1{
                margin: 10px 0px;
                margin-top: 0;
            }

            table{
                display: flex;
                flex-direction: row;
            }

            table, td, th {
                border-collapse: collapse;
            }
            
            thead {
                float: left;   
            }
            thead th {
                display: block;   
                text-align: left;
            }
            
            tbody {
                float: left;   
            }
            tbody tr {
                display: block;
                float: left;
            }
            tbody td {
                display: block;
            }
        }
    }

    
}