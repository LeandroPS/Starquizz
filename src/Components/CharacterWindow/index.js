import React, {Component} from 'react';
import PropTypes from 'prop-types';

import './styles.scss';

class CharacterWindow extends Component{

    constructor(){
        super();

        this.state = {
            show: false
        }
    }

    componentDidMount(){
        this.setState({show: true});
    }

    componentDidUpdate(prevProps) {
        if (prevProps.isMounted && !this.props.isMounted) {
          setTimeout(
            () => this.setState({ show: false }),
            300
          );
        } else if (!prevProps.isMounted && this.props.isMounted) {
          this.setState({ show: true });
        }
      }

    render(){
        const { character, onClose, getResource } = this.props;
        const { show } = this.state;

        return(
            <div className={`character-window ${show&&'show'}`}>
                <button className="close" onClick={onClose}>
                    <i className="fas fa-times"></i>
                </button>
                <div className="inner">
                    <div className="thumbnail" style={{backgroundImage: `url(${character.image})`}}>
    
                    </div>
                    <div className="details">
                        <h1>Details</h1>
                        <table>
                            <thead>
                                <tr>
                                    <th>Specie:</th>
                                    <th>Height:</th>
                                    <th>Planet:</th>
                                    <th>Movies:</th>
                                    <th>Vehicles:</th>
                                </tr>
                            </thead>
                            <tbody>
                                <tr>
                                    <td>{character.species.map((planet, i)=>{
                                        return `${i===0?'':', '}${getResource(planet)}`;
                                    })}</td>
                                    <td>
                                        {character.height} cm
                                    </td>
                                    <td>
                                        {getResource(character.homeworld)};
                                    </td>
                                    <td>
                                        {character.films.map((film, i)=>{
                                            return `${i===0?'':', '}${getResource(film)}`;
                                        })}
                                    </td>
                                    <td>
                                        {character.vehicles.map((vehicle, i)=>{
                                            return `${i===0?'':', '}${getResource(vehicle)}`;
                                        })}
                                    </td>
                                </tr>
                            </tbody>
                        </table>
                    </div>
                </div>
            </div>
        );
    }
}

CharacterWindow.propTypes = {
    character: PropTypes.object.isRequired, 
    getResource: PropTypes.func.isRequired,
    onClose: PropTypes.func.isRequired,
};

export default CharacterWindow;
