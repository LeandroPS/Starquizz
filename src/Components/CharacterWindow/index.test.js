import React from 'react';
import ReactDOM from 'react-dom';
import { create } from "react-test-renderer";
import { shallow } from 'enzyme';
import CharacterWindow from './';

describe('Character window', () => {
    const props = {
        character: {
            species: [],
            name: 'Luke Skywalker',
            height: 123,
            homeworld: 'earth',
            films: [],
            vehicles: []
        }, 
        getResource: ()=>{ return undefined},
        onClose: jest.fn()
    }

    it('renders without crashing', () => {
        const div = document.createElement('div');
        ReactDOM.render(<CharacterWindow {...props}/>, div);
        ReactDOM.unmountComponentAtNode(div);
    });

    it('renders correctly', () => {
        const tree = create(<CharacterWindow {...props}/>)
          .toJSON();
        expect(tree).toMatchSnapshot();
    });

    it('contain right info', () => {
        const component = shallow(<CharacterWindow {...props}/>);

        expect(component.find('table').text()).toContain('123 cm');
    });

    it('executes onClose function when click on close button', () => {
        const component = shallow(<CharacterWindow {...props}/>);
        component.find('button.close').simulate('click');
        expect(props.onClose).toHaveBeenCalled();
    });

});