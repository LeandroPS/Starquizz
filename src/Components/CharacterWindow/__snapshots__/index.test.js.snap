// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Character window renders correctly 1`] = `
<div
  className="character-window show"
>
  <button
    className="close"
    onClick={[MockFunction]}
  >
    <i
      className="fas fa-times"
    />
  </button>
  <div
    className="inner"
  >
    <div
      className="thumbnail"
      style={
        Object {
          "backgroundImage": "url(undefined)",
        }
      }
    />
    <div
      className="details"
    >
      <h1>
        Details
      </h1>
      <table>
        <thead>
          <tr>
            <th>
              Specie:
            </th>
            <th>
              Height:
            </th>
            <th>
              Planet:
            </th>
            <th>
              Movies:
            </th>
            <th>
              Vehicles:
            </th>
          </tr>
        </thead>
        <tbody>
          <tr>
            <td />
            <td>
              123
               cm
            </td>
            <td>
              ;
            </td>
            <td />
            <td />
          </tr>
        </tbody>
      </table>
    </div>
  </div>
</div>
`;
